### ISSUE TITLE
ACK storm

### ISSUE THREAD

**Issue by kenmcmil** [private-octopus/picoquic#543]
The following log shows a case where the picoquicdemo client sends ACK-only packets in response to ACK-only packets. For example, sent handshake packets 2, 3 and 4 have only ACK frames, and the only newly acknowledged packets are ACK-only. From quic-transport-draft-18:

```
13.1.1.  Sending ACK Frames

   An endpoint MUST NOT send more than one packet containing only an ACK
   frame per received packet that contains frames other than ACK and
   PADDING frames.  An endpoint MUST NOT send a packet containing only
   an ACK frame in response to a packet containing only ACK or PADDING
   frames, even if there are packet gaps which precede the received
   packet.  This prevents an indefinite feedback loop of ACKs.  The
   endpoint MUST however acknowledge packets containing only ACK or
   PADDING frames when sending ACK frames in response to other packets.
```

Also interesting about this log is that the client is blocked by the max_streams_bidi parameter, but does not send STREAM_BLOCKED. I don't think this is required, but it seems like a good idea to send it.

Client log follows:

```
Starting PicoQUIC connection to server IP = localhost, port = 4443
No root crt list specified, certificate will not be verified.
98ccdf4414b0f370: Sending packet type: 2 (initial), S0, Version ff000012,
98ccdf4414b0f370:     <98ccdf4414b0f370>, <3705aef4e7501563>, Seq: 0, pl: 1227
98ccdf4414b0f370:     Token length: 0
98ccdf4414b0f370:     Prepared 1207 bytes
98ccdf4414b0f370:     Crypto HS frame, offset 0, length 305: 0100012d03035f62...
98ccdf4414b0f370:     padding, 898 bytes

98ccdf4414b0f370: Sending 1252 bytes to 127.0.0.1:4443 at T=0.000000 (584b871eea96a)
Select returns 1232, from length 16
Local address updated
98ccdf4414b0f370: Receiving 1232 bytes from 127.0.0.1:4443 at T=0.638251 (584b871f86695)
98ccdf4414b0f370: Receiving packet type: 2 (initial), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 0, pl: 1207
98ccdf4414b0f370:     Token length: 0
98ccdf4414b0f370:     Decrypted 1190 bytes
98ccdf4414b0f370:     Crypto HS frame, offset 0, length 155: 020000970303c60a...
98ccdf4414b0f370:     padding, 1031 bytes

Processed 1232 bytes, state = 7 (client_handshake_start), return 0

98ccdf4414b0f370: Sending packet type: 2 (initial), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 1, pl: 1227
98ccdf4414b0f370:     Token length: 0
98ccdf4414b0f370:     Prepared 1207 bytes
98ccdf4414b0f370:     ACK (nb=0), 0
98ccdf4414b0f370:     padding, 1202 bytes

98ccdf4414b0f370: Sending 1252 bytes to 127.0.0.1:4443 at T=0.638548 (584b871f867be)
Select returns 1232, from length 16
98ccdf4414b0f370: Receiving 1232 bytes from 127.0.0.1:4443 at T=0.882222 (584b871fc1f98)
98ccdf4414b0f370: Receiving packet type: 2 (initial), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 1, pl: 1207
98ccdf4414b0f370:     Token length: 0
98ccdf4414b0f370:     Decrypted 1190 bytes
98ccdf4414b0f370:     ACK (nb=0), 0
98ccdf4414b0f370:     padding, 1185 bytes

Processed 1232 bytes, state = 7 (client_handshake_start), return 0

Select returns 1407, from length 16
98ccdf4414b0f370: Receiving 1407 bytes from 127.0.0.1:4443 at T=1.850277 (584b8720ae50f)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 0, pl: 1383
98ccdf4414b0f370:     Decrypted 1366 bytes
98ccdf4414b0f370:     Crypto HS frame, offset 0, length 1346: 0800003d003b0000...
98ccdf4414b0f370:     padding, 16 bytes

Processed 1407 bytes, state = 10 (client_almost_ready), return 0

Almost ready!

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 0, pl: 80
98ccdf4414b0f370:     Prepared 60 bytes
98ccdf4414b0f370:     ACK (nb=0), 0
98ccdf4414b0f370:     Crypto HS frame, offset 0, length 52: 14000030e2c98ea0...

98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <000000000000000d>, Seq: 0 (0), Phi: 0,
98ccdf4414b0f370:     Prepared 216 bytes
98ccdf4414b0f370:     NEW CONNECTION ID[1]: 0xd5561ed8ce820b00, a9da8b4a5cdb8676e9bcbb22ac774fee
98ccdf4414b0f370:     NEW CONNECTION ID[2]: 0xc1569479bc5f8063, 43d7aa012113cd3843df4b74c97b06ff
98ccdf4414b0f370:     NEW CONNECTION ID[3]: 0xd0869a3fff64e6dd, 6710ef0be22ef99cb610ab6443018417
98ccdf4414b0f370:     NEW CONNECTION ID[4]: 0x0a772fa9dea2bbf0, ada87cf386eb7f2a253e424fa6b0ea4a
98ccdf4414b0f370:     NEW CONNECTION ID[5]: 0x40936e81f9a59304, e3321f540465fc7b36a4d4b4dc34c847
98ccdf4414b0f370:     NEW CONNECTION ID[6]: 0xde48cd9a6927bdf1, 84af9098cab4c51f2b1d6d612c573f22
98ccdf4414b0f370:     NEW CONNECTION ID[7]: 0x967e7fffa8cf7680, cdc1ef067ab1717405b96cf6e8e8d469
98ccdf4414b0f370:     NEW CONNECTION ID[8]: 0xaa39dca378d368d5, 1acdd6cb65ff1f468922fdecd19c4070

98ccdf4414b0f370: Sending 349 bytes to 127.0.0.1:4443 at T=1.850472 (584b8720ae5d2)
Received SNI: localhost
ALPN not received.
Received transport parameter TLS extension (51 bytes):
Version: ff000012
    Supported version (8 bytes):
        ff000012
        49565931
    Extension list (36 bytes):
        Extension type: 5, length 2 (0x0005 / 0x0002), 6000
        Extension type: 4, length 4 (0x0004 / 0x0004), 80004000
        Extension type: 1, length 1 (0x0001 / 0x0001), 3c
        Extension type: 6, length 2 (0x0006 / 0x0002), 6000
        Extension type: 7, length 2 (0x0007 / 0x0002), 6000
        Extension type: 8, length 1 (0x0008 / 0x0001), 02

Connection established. Version = ff000012, I-CID: 98ccdf4414b0f370
Opening stream 0 to GET /index.html
98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <000000000000000d>, Seq: 1 (1), Phi: 0,
98ccdf4414b0f370:     Prepared 24 bytes
98ccdf4414b0f370:     Stream 0, offset 0, length 17, fin = 1: 474554202f696e64...
98ccdf4414b0f370:     padding, 4 bytes

98ccdf4414b0f370: Sending 53 bytes to 127.0.0.1:4443 at T=1.850675 (584b8720ae69d)
98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <000000000000000d>, Seq: 2 (2), Phi: 0,
98ccdf4414b0f370:     Prepared 1411 bytes
98ccdf4414b0f370:     ping, 1 bytes
98ccdf4414b0f370:     padding, 1410 bytes

98ccdf4414b0f370: Sending 1440 bytes to 127.0.0.1:4443 at T=1.850871 (584b8720ae761)
Select returns 1232, from length 16
98ccdf4414b0f370: Receiving 1232 bytes from 127.0.0.1:4443 at T=2.572071 (584b87215e891)
98ccdf4414b0f370: Receiving packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <3705aef4e7501563>, Seq: 0 (0), Phi: 0,
98ccdf4414b0f370:     Decrypted 1206 bytes
98ccdf4414b0f370:     ACK (nb=0), 0
98ccdf4414b0f370:     ACK (nb=0), 1
98ccdf4414b0f370:     ACK (nb=0), 2
98ccdf4414b0f370:     padding, 1191 bytes

Processed 1232 bytes, state = 14 (ready), return 0

Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=2.784811 (584b872192795)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 4, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 0
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 1, pl: 27
98ccdf4414b0f370:     Prepared 7 bytes
98ccdf4414b0f370:     ACK (nb=1), 4, 0

98ccdf4414b0f370: Sending 51 bytes to 127.0.0.1:4443 at T=2.784865 (584b8721927cb)
Select returns 50, from length 16
98ccdf4414b0f370: Receiving 50 bytes from 127.0.0.1:4443 at T=2.957235 (584b8721bc91d)
98ccdf4414b0f370: Receiving packet type: 6 (1rtt protected), S1,
98ccdf4414b0f370:     <3705aef4e7501563>, Seq: 12 (12), Phi: 0,
98ccdf4414b0f370:     Decrypted 24 bytes
98ccdf4414b0f370:     Stream 0, offset 0, length 4, fin = 1: 666f6f0a
98ccdf4414b0f370:     padding, 16 bytes

Stream 0 ended after 4 bytes
Opening stream 4 to GET /test.html
Opening stream 8 to GET /1234567
Opening stream 12 to GET /main.jpg
Opening stream 16 to GET /war-and-peace.txt
Opening stream 20 to GET /en/latest/
Opening stream 24 to GET /file-123K
Processed 50 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <000000000000000d>, Seq: 3 (3), Phi: 0,
98ccdf4414b0f370:     Prepared 26 bytes
98ccdf4414b0f370:     ACK (nb=1), 12, 0
98ccdf4414b0f370:     Stream 4, offset 0, length 16, fin = 1: 474554202f746573...

98ccdf4414b0f370: Sending 55 bytes to 127.0.0.1:4443 at T=2.957475 (584b8721bca0d)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=3.318686 (584b872214d08)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 10, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 1
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 2, pl: 29
98ccdf4414b0f370:     Prepared 9 bytes
98ccdf4414b0f370:     ACK (nb=2), 16, 4, 0

98ccdf4414b0f370: Sending 53 bytes to 127.0.0.1:4443 at T=3.318754 (584b872214d4c)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=3.587149 (584b8722565b7)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 2, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 2
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 3, pl: 34
98ccdf4414b0f370:     Prepared 14 bytes
98ccdf4414b0f370:     ACK (nb=3), 16, 4, 2, 0

98ccdf4414b0f370: Sending 58 bytes to 127.0.0.1:4443 at T=3.587203 (584b8722565ed)
Select returns 1232, from length 16
98ccdf4414b0f370: Receiving 1232 bytes from 127.0.0.1:4443 at T=3.865055 (584b87229a349)
98ccdf4414b0f370: Receiving packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <3705aef4e7501563>, Seq: 1 (1), Phi: 0,
98ccdf4414b0f370:     Decrypted 1206 bytes
98ccdf4414b0f370:     ACK (nb=0), 3
98ccdf4414b0f370:     padding, 1201 bytes

Processed 1232 bytes, state = 14 (ready), return 0

Select returns 1232, from length 16
98ccdf4414b0f370: Receiving 1232 bytes from 127.0.0.1:4443 at T=4.445371 (584b872327e25)
98ccdf4414b0f370: Receiving packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <3705aef4e7501563>, Seq: 4 (4), Phi: 0,
98ccdf4414b0f370:     Decrypted 1206 bytes
98ccdf4414b0f370:     Stream 4, offset 0, length 4, fin = 1: 666f6f0a
98ccdf4414b0f370:     padding, 1198 bytes

Stream 4 ended after 4 bytes
Processed 1232 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S1,
98ccdf4414b0f370:     <000000000000000d>, Seq: 4 (4), Phi: 0,
98ccdf4414b0f370:     Prepared 24 bytes
98ccdf4414b0f370:     ACK (nb=2), 12, 4, 1
98ccdf4414b0f370:     padding, 12 bytes

98ccdf4414b0f370: Sending 53 bytes to 127.0.0.1:4443 at T=4.445676 (584b872327f56)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=4.771494 (584b872377810)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 8, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 3
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 4, pl: 36
98ccdf4414b0f370:     Prepared 16 bytes
98ccdf4414b0f370:     ACK (nb=4), 16, 8, 4, 2, 0

98ccdf4414b0f370: Sending 60 bytes to 127.0.0.1:4443 at T=4.771551 (584b872377849)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=5.157250 (584b8723d5aec)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 5, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 4
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 5, pl: 36
98ccdf4414b0f370:     Prepared 16 bytes
98ccdf4414b0f370:     ACK (nb=4), 16, 8, 4-5, 2, 0

98ccdf4414b0f370: Sending 60 bytes to 127.0.0.1:4443 at T=5.157303 (584b8723d5b21)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=5.504444 (584b87242a726)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 12, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 5
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 6, pl: 35
98ccdf4414b0f370:     Prepared 15 bytes
98ccdf4414b0f370:     ACK (nb=5), 18, 16, 8, 4-5, 2, 0

98ccdf4414b0f370: Sending 59 bytes to 127.0.0.1:4443 at T=5.504494 (584b87242a758)
Select returns 47, from length 16
98ccdf4414b0f370: Receiving 47 bytes from 127.0.0.1:4443 at T=5.930506 (584b872492774)
98ccdf4414b0f370: Receiving packet type: 6 (1rtt protected), S1,
98ccdf4414b0f370:     <3705aef4e7501563>, Seq: 2 (2), Phi: 0,
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 4
98ccdf4414b0f370:     padding, 16 bytes

Processed 47 bytes, state = 14 (ready), return 0

Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=6.431009 (584b87250ca8b)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 1, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 6
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 7, pl: 36
98ccdf4414b0f370:     Prepared 16 bytes
98ccdf4414b0f370:     ACK (nb=4), 18, 16, 8, 4-5, 0-2

98ccdf4414b0f370: Sending 60 bytes to 127.0.0.1:4443 at T=6.431072 (584b87250caca)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=6.932987 (584b872587365)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 11, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 7
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 8, pl: 34
98ccdf4414b0f370:     Prepared 14 bytes
98ccdf4414b0f370:     ACK (nb=3), 16-18, 8, 4-5, 0-2

98ccdf4414b0f370: Sending 58 bytes to 127.0.0.1:4443 at T=6.933053 (584b8725873a7)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=7.491396 (584b87260f8ae)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 3, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 8
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 9, pl: 32
98ccdf4414b0f370:     Prepared 12 bytes
98ccdf4414b0f370:     ACK (nb=2), 16-18, 8, 0-5

98ccdf4414b0f370: Sending 56 bytes to 127.0.0.1:4443 at T=7.491494 (584b87260f910)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=8.064881 (584b87269b8db)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 6, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 9
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: a, pl: 32
98ccdf4414b0f370:     Prepared 12 bytes
98ccdf4414b0f370:     ACK (nb=2), 16-18, 8, 0-6

98ccdf4414b0f370: Sending 56 bytes to 127.0.0.1:4443 at T=8.064929 (584b87269b90b)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=8.712904 (584b872739c32)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: a, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 10
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: b, pl: 34
98ccdf4414b0f370:     Prepared 14 bytes
98ccdf4414b0f370:     ACK (nb=3), 16-18, 10, 8, 0-6

98ccdf4414b0f370: Sending 58 bytes to 127.0.0.1:4443 at T=8.712958 (584b872739c68)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=9.467538 (584b8727f1ffc)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 9, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 11
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: c, pl: 32
98ccdf4414b0f370:     Prepared 12 bytes
98ccdf4414b0f370:     ACK (nb=2), 16-18, 8-10, 0-6

98ccdf4414b0f370: Sending 56 bytes to 127.0.0.1:4443 at T=9.467592 (584b8727f2032)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=10.178701 (584b87289f9f7)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 14, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 12
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: d, pl: 31
98ccdf4414b0f370:     Prepared 11 bytes
98ccdf4414b0f370:     ACK (nb=3), 20, 16-18, 8-10, 0-6

98ccdf4414b0f370: Sending 55 bytes to 127.0.0.1:4443 at T=10.178779 (584b87289fa45)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=10.886974 (584b87294c8a8)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 20, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 13
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: e, pl: 33
98ccdf4414b0f370:     Prepared 13 bytes
98ccdf4414b0f370:     ACK (nb=4), 32, 20, 16-18, 8-10, 0-6

98ccdf4414b0f370: Sending 57 bytes to 127.0.0.1:4443 at T=10.887024 (584b87294c8da)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=11.708770 (584b872a152cc)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 22, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 14
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: f, pl: 35
98ccdf4414b0f370:     Prepared 15 bytes
98ccdf4414b0f370:     ACK (nb=5), 34, 32, 20, 16-18, 8-10, 0-6

98ccdf4414b0f370: Sending 59 bytes to 127.0.0.1:4443 at T=11.708833 (584b872a1530b)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=12.837632 (584b872b28c6a)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 30, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 15
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 10, pl: 37
98ccdf4414b0f370:     Prepared 17 bytes
98ccdf4414b0f370:     ACK (nb=6), 48, 34, 32, 20, 16-18, 8-10, 0-6

98ccdf4414b0f370: Sending 61 bytes to 127.0.0.1:4443 at T=12.837693 (584b872b28ca7)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=13.754462 (584b872c089c8)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 40, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 16
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

98ccdf4414b0f370: Sending packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <000000000000000d>, <3705aef4e7501563>, Seq: 11, pl: 40
98ccdf4414b0f370:     Prepared 20 bytes
98ccdf4414b0f370:     ACK (nb=7), 64, 48, 34, 32, 20, 16-18, 8-10, 0-6

98ccdf4414b0f370: Sending 64 bytes to 127.0.0.1:4443 at T=13.754507 (584b872c089f5)
Select returns 62, from length 16
98ccdf4414b0f370: Receiving 62 bytes from 127.0.0.1:4443 at T=14.733517 (584b872cf7a37)
98ccdf4414b0f370: Receiving packet type: 4 (handshake), S0, Version ff000012,
98ccdf4414b0f370:     <3705aef4e7501563>, <000000000000000d>, Seq: 42, pl: 38
98ccdf4414b0f370:     Decrypted 21 bytes
98ccdf4414b0f370:     ACK (nb=0), 17
98ccdf4414b0f370:     padding, 16 bytes

Processed 62 bytes, state = 14 (ready), return 0

No progress for 10 seconds. Closing. 
Received a request to close the connection.
98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <000000000000000d>, Seq: 5 (5), Phi: 0,
98ccdf4414b0f370:     Prepared 14 bytes
98ccdf4414b0f370:     ACK (nb=1), 12, 2
98ccdf4414b0f370:     application_close, Error 0x0000, Reason length 0

98ccdf4414b0f370: Sending 43 bytes to 127.0.0.1:4443 at T=14.733565 (584b872cf7a67)
Select returns 1232, from length 16
98ccdf4414b0f370: Receiving 1232 bytes from 127.0.0.1:4443 at T=15.857695 (584b872e0a189)
98ccdf4414b0f370: Receiving packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <3705aef4e7501563>, Seq: 18 (18), Phi: 0,
98ccdf4414b0f370:     Decrypted 1206 bytes
98ccdf4414b0f370:     ACK (nb=0), 5
98ccdf4414b0f370:     padding, 1201 bytes

Processed 1232 bytes, state = 17 (closing), return 0

98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <000000000000000d>, Seq: 6 (6), Phi: 0,
98ccdf4414b0f370:     Prepared 4 bytes
98ccdf4414b0f370:     application_close, Error 0x0000, Reason length 0

98ccdf4414b0f370: Sending 33 bytes to 127.0.0.1:4443 at T=15.857773 (584b872e0a1d7)
Select returns 1232, from length 16
98ccdf4414b0f370: Receiving 1232 bytes from 127.0.0.1:4443 at T=16.954390 (584b872f15d80)
98ccdf4414b0f370: Receiving packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <3705aef4e7501563>, Seq: 22 (22), Phi: 0,
98ccdf4414b0f370:     Decrypted 1206 bytes
98ccdf4414b0f370:     ACK (nb=0), 6
98ccdf4414b0f370:     padding, 1201 bytes

Processed 1232 bytes, state = 17 (closing), return 0

98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <000000000000000d>, Seq: 7 (7), Phi: 0,
98ccdf4414b0f370:     Prepared 4 bytes
98ccdf4414b0f370:     application_close, Error 0x0000, Reason length 0

98ccdf4414b0f370: Sending 33 bytes to 127.0.0.1:4443 at T=16.954440 (584b872f15db2)
Select returns 1232, from length 16
98ccdf4414b0f370: Receiving 1232 bytes from 127.0.0.1:4443 at T=18.184889 (584b873042423)
98ccdf4414b0f370: Receiving packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <3705aef4e7501563>, Seq: 32 (32), Phi: 0,
98ccdf4414b0f370:     Decrypted 1206 bytes
98ccdf4414b0f370:     ACK (nb=0), 7
98ccdf4414b0f370:     padding, 1201 bytes

Processed 1232 bytes, state = 17 (closing), return 0

98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <000000000000000d>, Seq: 8 (8), Phi: 0,
98ccdf4414b0f370:     Prepared 4 bytes
98ccdf4414b0f370:     application_close, Error 0x0000, Reason length 0

98ccdf4414b0f370: Sending 33 bytes to 127.0.0.1:4443 at T=18.184941 (584b873042457)
Select returns 1232, from length 16
98ccdf4414b0f370: Receiving 1232 bytes from 127.0.0.1:4443 at T=19.400711 (584b87316b171)
98ccdf4414b0f370: Receiving packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <3705aef4e7501563>, Seq: 48 (48), Phi: 0,
98ccdf4414b0f370:     Decrypted 1205 bytes
98ccdf4414b0f370:     ACK (nb=0), 8
98ccdf4414b0f370:     padding, 1200 bytes

Processed 1232 bytes, state = 17 (closing), return 0

98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <000000000000000d>, Seq: 9 (9), Phi: 0,
98ccdf4414b0f370:     Prepared 4 bytes
98ccdf4414b0f370:     application_close, Error 0x0000, Reason length 0

98ccdf4414b0f370: Sending 33 bytes to 127.0.0.1:4443 at T=19.400764 (584b87316b1a6)
Select returns 1232, from length 16
98ccdf4414b0f370: Receiving 1232 bytes from 127.0.0.1:4443 at T=20.679642 (584b8732a3544)
98ccdf4414b0f370: Receiving packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <3705aef4e7501563>, Seq: 36 (36), Phi: 0,
98ccdf4414b0f370:     Decrypted 1206 bytes
98ccdf4414b0f370:     ACK (nb=0), 9
98ccdf4414b0f370:     padding, 1201 bytes

Processed 1232 bytes, state = 17 (closing), return 0

98ccdf4414b0f370: Sending packet type: 6 (1rtt protected), S0,
98ccdf4414b0f370:     <000000000000000d>, Seq: 10 (10), Phi: 0,
98ccdf4414b0f370:     Prepared 4 bytes
98ccdf4414b0f370:     application_close, Error 0x0000, Reason length 0

98ccdf4414b0f370: Sending 33 bytes to 127.0.0.1:4443 at T=20.679704 (584b8732a3582)
quic/picoquic/quicctx.c:839 [picoquic_delete_path]: delete path[8] (557dc960)

quic/picoquic/quicctx.c:839 [picoquic_delete_path]: delete path[7] (557dc5f0)

quic/picoquic/quicctx.c:839 [picoquic_delete_path]: delete path[6] (557dc280)

quic/picoquic/quicctx.c:839 [picoquic_delete_path]: delete path[5] (557dbf10)

quic/picoquic/quicctx.c:839 [picoquic_delete_path]: delete path[4] (557dbba0)

quic/picoquic/quicctx.c:839 [picoquic_delete_path]: delete path[3] (557db830)

quic/picoquic/quicctx.c:839 [picoquic_delete_path]: delete path[2] (557db4c0)

quic/picoquic/quicctx.c:839 [picoquic_delete_path]: delete path[1] (557d9440)

quic/picoquic/quicctx.c:839 [picoquic_delete_path]: delete path[0] (557d6670)

Client exit with code = 0

```

**Comment by huitema** [comment]
The complete prohibition against ack of ack is probably an error in the spec, and I will open an issue. Nodes actually need some ack-of-ack in order to clear the ack list. But even if we lift that requirement somewhat, the ping-pong behavior in this log is wrong. OTOH it takes two to ping pong, which is why I did not see it in practice. It may be tied to handshake mode. Will check.

There is something else at the end of the log: two many repeats of the app close packet in the closing state. Will check that too.

The blocked frames are optional. No protocol machine should depend on them.

Application profiles like H3 specify that servers should allow some number of uni and bidir streams to work. The blocked on bidir case is pretty much a server error. The current behavior is to wait until the window opens, and kill the connection if it doesn't after a short time. I believe that's correct.

**Comment by huitema** [comment]
Found it. Silly bug indeed. The connection has reached the point where it is "ready". At that point, picoquic does not care about the handshake packets. It just acks them in case the peer needs an ACK -- but there was a bug and it would ACK the ACK. Will be fixed soon.

**Comment by kenmcmil** [comment]
OK, by the way I also have a trace in which the client gets into an ACK storm with initial packets. If it's helpful, I'll post it.

**Comment by huitema** [comment]
@kenmcmil by all means, but please open a separate issue for that.
